api/repositories/org_repository.go

```diff
@@ -125,61 +129,105 @@ func (r *OrgRepo) CreateOrg(ctx context.Context, info authorization.Info, org Cr
-			},
+	var orgCR *workloads.CFOrg
+	orgCR, err = r.createOrgCR(ctx, info, userClient, &workloads.CFOrg{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      OrgPrefix + uuid.NewString(),
+			Namespace: r.rootNamespace,
 		},
-		OrgResourceType,
-	)
+		Spec: workloads.CFOrgSpec{
+			Name: org.Name,
+		},
+	})
+
 	if err != nil {
-		if webhookError, ok := webhooks.WebhookErrorToValidationError(err); ok {
+		if webhookError, ok := webhooks.WebhookErrorToValidationError(err); ok { // untested
 			return OrgRecord{}, apierrors.NewUnprocessableEntityError(err, webhookError.Error())
 		}
 		return OrgRecord{}, err
 	}
 
-	if err := setCascadingDelete(ctx, userClient, orgGUID); err != nil {
-		return OrgRecord{}, err
-	}
-
 	orgRecord := OrgRecord{
 		Name:        org.Name,
-		GUID:        anchor.Name,
+		GUID:        orgCR.Name,
 		Suspended:   org.Suspended,
 		Labels:      org.Labels,
 		Annotations: org.Annotations,
-		CreatedAt:   anchor.CreationTimestamp.Time,
-		UpdatedAt:   anchor.CreationTimestamp.Time,
+		CreatedAt:   orgCR.CreationTimestamp.Time,
+		UpdatedAt:   orgCR.CreationTimestamp.Time,
 	}
 
 	return orgRecord, nil
 }
 
-func setCascadingDelete(ctx context.Context, userClient client.Client, orgGUID string) error {
-	oldHC := v1alpha2.HierarchyConfiguration{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      hierarchyMetadataName,
-			Namespace: orgGUID,
-		},
+func (r *OrgRepo) createOrgCR(ctx context.Context,
+	info authorization.Info,
+	userClient client.Client,
+	org *workloads.CFOrg,
+) (*workloads.CFOrg, error) {
+	err := userClient.Create(ctx, org)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create cf org: %w", apierrors.FromK8sError(err, OrgResourceType))
 	}
-	newHC := oldHC
-	newHC.Spec.AllowCascadingDeletion = true
 
-	if err := userClient.Patch(ctx, &newHC, client.MergeFrom(&oldHC)); err != nil {
-		return fmt.Errorf("failed to update hierarchy configuration: %w", apierrors.FromK8sError(err, OrgResourceType))
+	timeoutCtx, cancelFn := context.WithTimeout(ctx, r.timeout)
+	defer cancelFn()
+
+	watch, err := r.privilegedClient.Watch(timeoutCtx, &workloads.CFOrgList{},
+		client.InNamespace(org.Namespace),
+		client.MatchingFields{"metadata.name": org.Name},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set up watch on cf org: %w", apierrors.FromK8sError(err, OrgResourceType))
 	}
 
-	return nil
+	conditionReady := false
+	var createdOrg *workloads.CFOrg
+	for res := range watch.ResultChan() {
+		var ok bool
+		createdOrg, ok = res.Object.(*workloads.CFOrg)
+		if !ok {
+			// should never happen, but avoids panic above
+			continue
+		}
+		if meta.IsStatusConditionTrue(createdOrg.Status.Conditions, StatusConditionReady) {
+			watch.Stop()
+			conditionReady = true
+			break
+		}
+	}
+
+	if !conditionReady {
+		return nil, fmt.Errorf("cf org did not get Condition `Ready`: 'True' within timeout period %d ms", r.timeout.Milliseconds())
+	}
+
+	// wait for the namespace to be created and user to have permissions
```

Comment:
Now that we have our custom orgs resource and custom reconciler, can we not update the org status to `ready` only when the underlying namespace is there and get rid of this code from the repo?

Reply:
For now we believe we still need the namespace access check. The CFOrg shouldn't become Ready until the namespace exists, but we don't have a good way to check if HNC has propagated the rolebindings into the new namespace. When we replace HNC, the new solution should allow us to track rolebinding propagation easily. We'll be able to remove this then.

Reply:
Yes, agree, my point was about moving those checks to the cforg reconciler that sets the ready flag whenever the very same conditions are met. Then the repo should only check the ready org flag, rather than checking the namespace as well. This would be beneficial to theoretical  kubectl users as well.

But I can also see your point that this is probably a temporary workaround..